swifttool/manager.py

```diff
@@ -0,0 +1,288 @@
+
+_STATUS_FILE = 'capacity.status'
+
+
+class CapacityManager(object):
+
+    def __init__(self, iterations=10,
+                 def_file='/etc/swift/ring_definition.yml',
+                 scaleup=True, metadata=None,
+                 outdir='/etc/swift/'):
+        self.iterations = int(iterations)
+        self.def_file = def_file
+        self.scaleup = scaleup
+        self.cluster_changes = {}
+        self.outdir = outdir
+        self.cluster_config = None
+        self.metadata = metadata
+        # Save the copy
+        shutil.copyfile(self.def_file, self.def_file + ".orig")
+        self.current_status = os.path.join(outdir, _STATUS_FILE)
+        self._write_status(0)
+        self._calculate_cluster_changes()
+
+    def __del__(self):
+        # Cleanup the files we create
+        os.remove(self.current_status)
+        os.remove(self.def_file + ".orig")
+
+    def update_cluster(self):
+        LOG.info("Starting to update the cluster")
+        min_part_hours = int(self._get_builder('account').min_part_hours)
+        for iteration in range(1, self.iterations + 1):
+            changed, cfg = self._update_cluster_nodes()
+            if not changed:
+                break
+            # Write up the new config in /etc/swift
+            with open(self.def_file, 'w') as stream:
+                yaml.dump(cfg, stream, default_flow_style=False,
+                          explicit_start=True)
+            # Update status
+            self._write_status(int((100 * iteration) / self.iterations))
+            LOG.info("Sleeping for %d hours 1 minutes" % min_part_hours)
+            # Sleep for min_part_hours + minute
+            time.sleep((min_part_hours * (60 * 60)) + 60)
+            # Check if rings need to be rebalanced
+            self._rebalance_rings(cfg, min_part_hours)
+            LOG.info("Iteration " + str(iteration) + " complete")
+
+        LOG.info("Finished updating the cluster")
+        # Things finished with no errors - copy the original file back
+        shutil.copyfile(self.def_file + ".orig", self.def_file)
+
+    def _rebalance_rings(self, config, min_part_hours):
+        # If the balance factor or rings is more than 5
+        # we force rebalance to assign the partitions
+        # This function will force reassignment and also
+        # updates the rings
+        # Note: This is usually rare and happens when too much capacity
+        # was being added in single iteration
+        LOG.debug("Checking if rings need to be rebalanced")
+        while True:
+            need_modify = False
+            for ring_type in RING_TYPES:
+                if need_modify:
+                    continue
+                builder = self._get_builder(ring_type)
+                balance = builder.get_balance()
+                if balance > 5 and balance / 100.0 > builder.overload:
+                    need_modify = True
+            # Did not modify rings, return
+            if not need_modify:
+                return
```

Comment:
you could also just `break` here; the function will implicitly return `None`, and this function always returns `None` anyway.  just a think.